corehq/blobs/s3db.py

```diff
@@ -141,3 +144,148 @@ def __enter__(self):
 
     def __exit__(self, *args):
         self.obj.close()
+
+
+class OpenFileOSUtils(object):
+
+    def get_file_size(self, fileobj):
+        if not hasattr(fileobj, 'fileno'):
+            pos = fileobj.tell()
+            try:
+                fileobj.seek(0, os.SEEK_END)
+                return fileobj.tell()
+            finally:
+                fileobj.seek(pos)
+        return os.fstat(fileobj.fileno()).st_size
+
+    def open_file_chunk_reader(self, fileobj, start_byte, size, callback):
+        full_size = self.get_file_size(fileobj)
+        return ReadOpenFileChunk(fileobj, start_byte, size, full_size,
+                                 callback, enable_callback=False)
+
+    def open(self, filename, mode):
+        raise NotImplementedError
+
+    def remove_file(self, filename):
+        raise NotImplementedError
+
+    def rename_file(self, current_filename, new_filename):
+        raise NotImplementedError
+
+
+class ReadOpenFileChunk(ReadFileChunk):
+    """Wrapper for OpenFileChunk that implements ReadFileChunk interface
+    """
+
+    def __init__(self, fileobj, start_byte, chunk_size, full_file_size, *args, **kw):
+
+        class FakeFile:
+            def seek(self, pos):
+                pass
+
+        length = min(chunk_size, full_file_size - start_byte)
+        self._chunk = OpenFileChunk(fileobj, start_byte, length)
+        super(ReadOpenFileChunk, self).__init__(
+            FakeFile(), start_byte, chunk_size, full_file_size, *args, **kw)
+        assert self._size == length, (self._size, length)
+
+    def __repr__(self):
+        return ("<ReadOpenFileChunk {} offset={} length={}>".format(
+            self._chunk.file,
+            self._start_byte,
+            self._size,
+        ))
+
+    def read(self, amount=None):
+        data = self._chunk.read(amount)
+        if self._callback is not None and self._callback_enabled:
+            self._callback(len(data))
+        return data
+
+    def seek(self, where):
+        old_pos = self._chunk.tell()
+        self._chunk.seek(where)
+        if self._callback is not None and self._callback_enabled:
+            # To also rewind the callback() for an accurate progress report
+            self._callback(where - old_pos)
+
+    def tell(self):
+        return self._chunk.tell()
+
+    def close(self):
+        self._chunk.close()
+        self._chunk = None
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self.close()
+
+
+class OpenFileChunk(object):
+    """A wrapper for reading from a file-like object from multiple threads
+
+    Each thread reading from the file-like object should have its own
+    private instance of this class.
+    """
```

Comment:
`S3Transfer` uses multiple threads to upload chunks of a file. Each thread is given a separate chunk of the file to upload, but the default implementation expects a file path instead of an open file object, which allows it to open a separate file handle for each thread/chunk. `OpenFileChunk` allows it to read from a single shared open file object.